auth/api/iam/api.go

```diff
@@ -897,3 +898,38 @@ func createOAuth2BaseURL(webDID did.DID) (*url.URL, error) {
 	result = result.JoinPath(basePath, "oauth2", webDID.String())
 	return result, err
 }
+
+// subjectExists checks whether the given subject is known on the local node.
+func (r Wrapper) subjectExists(ctx context.Context, subjectID string) error {
+	_, err := r.selectDID(ctx, subjectID)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// subjectExists checks whether the given subject is known on the local node.
+func (r Wrapper) subjectOwns(ctx context.Context, subjectID string, subjectDID did.DID) (bool, error) {
+	dids, err := r.subjectManager.List(ctx, subjectID)
+	if err != nil {
+		return false, err
+	}
+	for _, d := range dids {
+		if d.Equals(subjectDID) {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (r Wrapper) selectDID(ctx context.Context, subjectID string) (*did.DID, error) {
+	// TODO: List() should return the DIDs in preferred order?
+	dids, err := r.subjectManager.List(ctx, subjectID)
+	if err != nil {
+		return nil, err
+	}
+	if len(dids) == 0 {
+		return nil, core.NotFoundError("subject not found")
+	}
+	return &dids[0], nil
+}
```

Comment:
len check is indeed obsolete.

It's used in 3 places:
- OpenID4VCI, as `client_id` and to construct the redirect URL, and subsequently to sign the proof.
- OpenID4VP user flow, to construct the redirect URL

To be able to remove it, we need to change DID parameter to subject in the redirect URL first, but that's part of https://github.com/nuts-foundation/nuts-node/pull/3264. So I'm adding a TODO to that PR.
And OpenID4VCI, although unmaintained, could have some form of service layer so these concerns are better separated. But I don't want to spend too much time on it, since nobody is using it and it's unclear whether it will ever be used in its current state.

Reply:
https://github.com/nuts-foundation/nuts-node/pull/3317